### PR TITLE
[SPARK-10097] Adds `shouldMaximize` flag to `ml.evaluation.Evaluator`

### PR DESCRIPTION
**Pull request by feynmanliang**
Previously, users of evaluator (`CrossValidator` and `TrainValidationSplit`) would only maximize the metric in evaluator, leading to a hacky solution which negated metrics to be minimized and caused erroneous negative values to be reported to the user.

This PR adds a `isLargerBetter` attribute to the `Evaluator` base class, instructing users of `Evaluator` on whether the chosen metric should be maximized or minimized.

CC @jkbradley

### CODE REVIEW

**Comment by feynmanliang** [comment]
jenkins test this please

**Comment by jkbradley** [comment]
LGTM pending tests

**Comment by feynmanliang** [comment]
Jenkins test this please

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/Evaluator.scala [hunk 1]
```diff
@@ -46,5 +46,12 @@ abstract class Evaluator extends Params {
    */
   def evaluate(dataset: DataFrame): Double
 
+  /**
+   * Indicates whether the metric returned by [[evaluate()]] should be maximized (true)
+   * or minimized (false).
+   * A given evaluator may support multiple metrics which may be maximized or minimized.
+   */
+  def shouldMaximize: Boolean
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/Evaluator.scala]: * `Evaluator` reports the metrics but not maximizes them. I would suggest `isLargerBetter` instead.
* This is a break change if we don't put a default value. I would suggest a default value `true` to keep it consistent with implementations in Spark 1.4.

**Comment by mengxr** [comment]
@feynmanliang Could you change the title to `ml.evaluation.Evaluator`?

**Comment by jkbradley** [comment]
Merging with master and branch-1.5

**Pull request closed without merging**